File: c/algos_c/merge_sort/merge_sort.c
#include <stdlib.h>
#include <stdio.h>
#include <time.h>

void fillArray(int* ary, int aryLen, int randMin, int randMax)
{
	for (int i = 0; i < aryLen; i++) {
		ary[i] = randMin + (rand() % (randMax - randMin));
	}
}

void printArray(int* ary, int aryLen)
{
	for (int i = 0; i < aryLen; i++) {
		printf("%d ", ary[i]);
	}
	printf("\n");
}

void mergeSortRecurse(int* ary, int* copy, int left, int right)
{
	int size = right - left;
	int mid = left + (size / 2);
	if (size == 1) {
		return;
	}
	// if size > 2, we can recurse on our left/right parts
	if (size > 2) {
		mergeSortRecurse(ary, copy, left, mid);
		mergeSortRecurse(ary, copy, mid, right);
	}

	// copy the l/r parts in ascending order into a copy array
	int lbegin = left;
	int rbegin = mid;
	for (int i = 0; i < size; i++) {
		if (lbegin == mid) { // done with left
			copy[left + i] = ary[rbegin];
			rbegin++;
		} else if (rbegin == right) { // done with right
			copy[left + i] = ary[lbegin];
			lbegin++;
		} else {
			int lsmaller = ary[lbegin] < ary[rbegin];
			if (lsmaller) {
				copy[left + i] = ary[lbegin];
				lbegin++;
			} else {
				copy[left + i] = ary[rbegin];
				rbegin++;
			}
		}
	}

	// copy the sorted items back
	for (int i = 0; i < size; i++) {
		ary[left + i] = copy[left + i];
	}
}

void mergeSort(int* ary, int aryLen)
{
	int copy[aryLen];
	int* copy_ptr = &copy[0];
	for (int i = 0; i < aryLen; i++) {
		copy[i] = ary[i];
	}
	mergeSortRecurse(ary, copy_ptr, 0, aryLen);
}

int main()
{
	srand(time(NULL));

	int randMin = 10;
	int randMax = 100;
	int aryLen = 10;
	int ary[aryLen];
	int* ary_ptr = &ary[0];

	fillArray(ary_ptr, aryLen, randMin, randMax);
	printArray(ary_ptr, aryLen);
	mergeSort(ary_ptr, aryLen);
	printArray(ary_ptr, aryLen);


	return 0;
}

File: python/merge_sort/merge_sort.py
from random import shuffle

def mergeSortRecurse(ary, copy, left, right):
  size = right - left
  # can't sort only 1 element
  if (size == 1):
    return

  # size > 2, we can split in half and recurse onto both
  mid = (left + right) / 2
  if (size > 2):
    mergeSortRecurse(ary, copy, left, mid)
    mergeSortRecurse(ary, copy, mid, right)

  # implicit size >= 2, sort our halves into ascending order
  lbegin = left
  rbegin = mid
  i = 0
  # while both sections have items
  while (lbegin < mid and rbegin < right):
    if (ary[lbegin] < ary[rbegin]):
      copy[left + i] = ary[lbegin]
      lbegin += 1
    else:
      copy[left + i] = ary[rbegin]
      rbegin += 1
    i += 1
  # one of the sections is finished, copy the rest from the unfinished side
  while (lbegin < mid):
    copy[left + i] = ary[lbegin]
    lbegin += 1
    i += 1
  while (rbegin < right):
    copy[left + i] = ary[rbegin]
    rbegin += 1
    i += 1

  # write the values from ordered copy back into the original
  for i in xrange(size):
    ary[left + i] = copy[left + i]

def mergeSort(ary):
  aryLen = len(ary)
  copy = range(aryLen)
  mergeSortRecurse(ary, copy, 0, aryLen)

aryLen = 20
ary = range(aryLen)
shuffle(ary)
print ary
mergeSort(ary)
print ary